File: apply_dashboard_defaults.py
import os
import re

app_tsx_path = os.path.join("frontend", "src", "App.tsx")
with open(app_tsx_path, "r", encoding="utf-8") as f:
    content = f.read()

helper = """export const getDashboardSetting = (key: string, def: string) => {
  try {
    const d = JSON.parse(localStorage.getItem('iqassess_dashboard_config') || '{}');
    return d[key] || def;
  } catch(e) {
    return def;
  }
};

export default function App() {"""

if "export const getDashboardSetting" not in content:
    content = content.replace("export default function App() {", helper)

replacements = [
  ("omrInstitutionName", "institution"),
  ("mcqInstitution", "institution"),
  ("adsInstitution", "institution"),
  ("bpdsInstitution", "institution"),
  ("qpInstitutionName", "institution"),
  ("newRubricInstitution", "institution"),
  ("essayBuilderInstitution", "institution"),
  ("mcqBuilderInstitution", "institution"),
  
  ("baCourse", "course"),
  ("itemAnalysisCourse", "course"),
  ("currCourseName", "course"),
  ("bpdsCourse", "course"),
  ("newRubricCourse", "course"),
  ("essayBuilderCourse", "course"),
  ("mcqBuilderCourse", "course"),

  ("baSubject", "subject"),
  ("mcqSubjectTitle", "subject"),
  ("bpSubject", "subject"),
  ("adsSubject", "subject"),
  ("bpdsSubject", "subject"),
  ("qpSubject", "subject"),
  ("newRubricSubject", "subject"),
  ("essayBuilderSubject", "subject"),
  ("mcqBuilderSubject", "subject"),

  ("baTopic", "topic"),
  ("mcqTopicText", "topic"),
  ("mcqTopic", "topic"),
  ("adsTopic", "topic"),
  ("bpdsTopic", "topic"),
  ("essayBuilderTopic", "topic"),
  ("mcqBuilderTopic", "topic")
]

for stateName, key in replacements:
    setterName = "set" + stateName[0].upper() + stateName[1:]
    # Match const [state, setState] = useState<string>('val')
    # Use re.sub with a function to preserve the default value if it's not already wrapped
    pattern = r"const \[" + stateName + r", " + setterName + r"\] = useState(?:<string>)?\((.*?)\);"
    
    def replacer(match):
        p1 = match.group(1)
        if "getDashboardSetting" in p1 or "getInitialHistory" in p1:
            return match.group(0)
        return f"const [{stateName}, {setterName}] = useState<string>(getDashboardSetting('{key}', {p1}));"

    content = re.sub(pattern, replacer, content)

with open(app_tsx_path, "w", encoding="utf-8") as f:
    f.write(content)

print("Successfully replaced App.tsx state variables.")

File: apply_dashboard_defaults.js
const fs = require('fs');
const path = require('path');

const appTsxPath = path.join(__dirname, 'frontend', 'src', 'App.tsx');
let content = fs.readFileSync(appTsxPath, 'utf8');

// Inject helper
if (!content.includes('const getDashboardSetting')) {
  content = content.replace(
    'export default function App() {',
    'export const getDashboardSetting = (key: string, def: string) => {\\n' +
    '  try {\\n' +
    '    const d = JSON.parse(localStorage.getItem(\\'iqassess_dashboard_config\\') || \\'{}\\');\\n' +
    '    return d[key] || def;\\n' +
    '  } catch(e) {\\n' +
    '    return def;\\n' +
    '  }\\n' +
    '};\\n\\n' +
    'export default function App() {'
  );
}

const replacements = [
  { stateName: 'omrInstitutionName', key: 'institution' },
  { stateName: 'mcqInstitution', key: 'institution' },
  { stateName: 'adsInstitution', key: 'institution' },
  { stateName: 'bpdsInstitution', key: 'institution' },
  { stateName: 'qpInstitutionName', key: 'institution' },
  { stateName: 'newRubricInstitution', key: 'institution' },
  { stateName: 'essayBuilderInstitution', key: 'institution' },
  { stateName: 'mcqBuilderInstitution', key: 'institution' },
  
  { stateName: 'baCourse', key: 'course' },
  { stateName: 'itemAnalysisCourse', key: 'course' },
  { stateName: 'currCourseName', key: 'course' },
  { stateName: 'bpdsCourse', key: 'course' },
  { stateName: 'newRubricCourse', key: 'course' },
  { stateName: 'essayBuilderCourse', key: 'course' },
  { stateName: 'mcqBuilderCourse', key: 'course' },

  { stateName: 'baSubject', key: 'subject' },
  { stateName: 'mcqSubjectTitle', key: 'subject' },
  { stateName: 'bpSubject', key: 'subject' },
  { stateName: 'adsSubject', key: 'subject' },
  { stateName: 'bpdsSubject', key: 'subject' },
  { stateName: 'qpSubject', key: 'subject' },
  { stateName: 'newRubricSubject', key: 'subject' },
  { stateName: 'essayBuilderSubject', key: 'subject' },
  { stateName: 'mcqBuilderSubject', key: 'subject' },

  { stateName: 'baTopic', key: 'topic' },
  { stateName: 'mcqTopicText', key: 'topic' },
  { stateName: 'mcqTopic', key: 'topic' },
  { stateName: 'adsTopic', key: 'topic' },
  { stateName: 'bpdsTopic', key: 'topic' },
  { stateName: 'essayBuilderTopic', key: 'topic' },
  { stateName: 'mcqBuilderTopic', key: 'topic' }
];

replacements.forEach(({ stateName, key }) => {
  const setterName = 'set' + stateName.charAt(0).toUpperCase() + stateName.slice(1);
  const regexStr = 'const \\\\[\\\\s*' + stateName + '\\\\s*,\\\\s*' + setterName + '\\\\s*\\\\]\\\\s*=\\\\s*useState(?:<string>)?\\\\((.*?)\\\\)';
  const regex = new RegExp(regexStr, 'g');
  
  content = content.replace(regex, (match, p1) => {
    if (p1.includes('getDashboardSetting') || p1.includes('getInitialHistory')) {
      return match;
    }
    return 'const [' + stateName + ', ' + setterName + '] = useState<string>(getDashboardSetting(\\'' + key + '\\', ' + p1 + '))';
  });
});

fs.writeFileSync(appTsxPath, content);
console.log('App.tsx updated with getDashboardSetting helpers.');
